### PR TITLE
Provide field validation in SparkApplication CRD yaml

### LINKED ISSUE #364: Provide complete field validation in CRD
The following fields are defined in `SparkApplicationSpec` with nested structs but not in `CustomResourceValidation`:
```
Image
InitContainerImage
ImagePullPolicy
MainClass
MainApplicationFile
Arguments
SparkConf
HadoopConf
SparkConfigMap
HadoopConfigMap
Volumes
NodeSelector
MemoryOverheadFactor
```

This may lead to ambigous and uninformative error messages during application of `CustomResource` as well as feedbackless wrong job configuration.

- [x] validation in yaml
- [ ] validation in `CustomResourceValidation` inside operator

fyi @sarjeet2013 @liyinan926

### PR DESCRIPTION
**Pull request by amukhopad**
Partially fixes #364

@liyinan926 @sarjeet2013

### CODE REVIEW

File: manifest/spark-operator-crds.yaml [hunk 1]
```diff
@@ -100,6 +148,14 @@ spec:
...
               - Python
               - R
+            sparkVersion:
+              type: string
+            memoryOverheadFactor:
+              type: string
+          required:
+          - type
+          - sparkVersion
+          - mode
```
[the remark this answers is not in the document]
Comment (amukhopad) [reply manifest/spark-operator-crds.yaml]: wasn't client mode support added in 2.4? as far as I tested it works

https://issues.apache.org/jira/browse/SPARK-23146

[the remark this answers is not in the document]
Comment (liyinan926) [reply manifest/spark-operator-crds.yaml]: Yes, `client` mode was added in 2.4, but the operator doesn't really support and it probably doesn't even make sense for the operator to support it.

File: manifest/spark-operator-crds.yaml [hunk 2]
```diff
@@ -59,12 +95,24 @@ spec:
                 instances:
                   minimum: 1
                   type: integer
+            nodeSelector:
+              type: object
+            falureRetries:
```
[the remark this answers is not in the document]
Comment (amukhopad) [reply manifest/spark-operator-crds.yaml]: @liyinan926 fixed.

File: manifest/spark-operator-crds.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (amukhopad) [reply manifest/spark-operator-crds.yaml]: Removed from required.

[the remark this answers is not in the document]
Comment (sarjeet2013) [reply manifest/spark-operator-crds.yaml]: @liyinan926 Agree with the `Client` mode support. We had internal discussion and planned to not support it as well. It doesn't make much sense here with Spark on k8s.

**Comment by amukhopad** [comment]
@liyinan926 

> There's a command-line tool for generating CRD including jsonschema validation at https://github.com/kubernetes-sigs/controller-tools/blob/master/cmd/crd/cmd/generate.go. We might want to use that.

Does it mean this definition will be removed eventually?

https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/pkg/crd/sparkapplication/crd.go#L59

**Comment by liyinan926** [comment]
Yes, we will deprecate runtime CRD creation within the operator and switch to using generated CRDs eventually.

**Pull request merged**